tests/ops/functions/test_equal.py

```diff
@@ -161,6 +161,80 @@
 )
 
 
+equal_hamiltonians = [
```

Comment:
I'm a bit confused: the PR mentions adding support for observables but also mentions that `qml.Hermitian` worked before. Is the addition specific to Hamiltonian and Tensor? Is it going to be compatible with Prod?

If general observable support is being added, then the following edge cases would be great to be checked:
* Elementary observable classes: `Projector`, `SparseHamiltonian`, `Hermitian`.

This is on in addition to the fact that the existing cases are already great!

Reply:
Good question. This is adding the logic from `Hamiltonian.compare` to the equals function, which allows Hamiltonians to be compared to other Hamiltonians, and Hamiltonians to be compared to Tensors. 

Once it's not attached to a `qml.Hamiltonian`, the same logic will also compare two Tensors to each other, which we couldn't do previously, and it will compare two Hermitians, but not as well as the original `qml.equal` function. 

`Observables` is misleading, I'll change it to `Hamiltonians` and `Tensors`.

Reply:
`Prod` is not an `Observable` object and doesn't have the `_obs_data` attribute that this uses, so this is not compatible with `Prod`. `Prod` will need new logic completely separate from `Tensor`.

It does not work for `Projector` or `SparseHamiltonian` for a similar reasons - while they are Observables and technically have an `_obs_data` attribute, calling it raises an error.

Reply:
Also, this logic is one of many places where things only work if the `Tensors` are really tensors. This is also true in `Hamiltonian.compare`. I will add that to the drawbacks, see above for details.

Reply:
:+1: